keps/sig-node/2400-node-swap/README.md

```diff
@@ -0,0 +1,727 @@
+
+This user story is addressed by scenarios 1 and 2, and could benefit from 3.
+
+#### Low footprint systems
+
+For example, edge devices with limited memory.
+
+- Edge compute systems/devices with small memory footprints (\<2Gi) https://github.com/kubernetes/kubernetes/issues/53533#issuecomment-751398086
+- Clusters with nodes \<4Gi memory https://github.com/kubernetes/kubernetes/issues/53533#issuecomment-751404417
+
+This user story is addressed by scenario 2, and could benefit from 3.
+
+#### Virtualization management overhead
+
+This would apply to virtualized Kubernetes workloads such as VMs launched by kubevirt.
+
+Every VM comes with a management related overhead which can sporadically be pretty significant (memory streaming, SRIOV attachment, gpu attachment, virtio-fs, …). Swap helps to not request much more memory to deal with short term worst-case scenarios.
+
+With virtualization, clusters are typically provisioned based on the workloads’ memory consumption, and any infrastructure container overhead is overcommitted. This overhead could be safely swapped out.
+
+- Required for live migration of VMs https://github.com/kubernetes/kubernetes/issues/53533#issuecomment-754878431
+
+This user story is addressed by scenario 2, and could benefit from 3.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+Having swap available on a system reduces predictability. When swap is available to workloads, and is not accounted for on an individual workload-by-workload basis
+
+First, this risk is mitigated by preventing any workloads from using swap by default, even if it is enabled on a system. This will allow a cluster administrator to test swap utilization just at the system level without introducing unpredictability to workload resource utilization.
+
+Additionally, we mitigate this risk by quantifying system stability and then gathering test and production data to determine if system stability remains the same or is improved when swap is available to the system and/or workloads.
+
+Since swap provisioning is out of scope of this proposal, this enhancement poses little risk to Kubernetes clusters that will not enable swap.
+
+## Design Details
+
+### TL;DR
+
+In a nutshell, the following implementation are planned for Memory Swap Support
+in 1.22 GKE alpha
+
+1. Having a feature gate `SupportNodeMemorySwap` guarding against the memory
+   swap support feature
+2. Keep the default value of kubelet flag `--fail-on-swap` to `true` in order
+   to minimize the blast radius
+3. Introducing two new kubelet config `MemorySwapLimit` and `Swappiness`
+4. Introducing two new CRI parameter `memory_swap_limit_in_bytes` and `memory_swappiness`
```

Comment:
From what I can see runc does not support setting swappiness currently, and we can just use the system defaults. @karan had previously suggested that we leave swappiness to a cluster admin to configure, e.g. with a daemonset.

Runc appears to only support `memory-swap`: https://github.com/opencontainers/runc/blob/c2c35aefee6b2822f4fa9412e08afa4f7798a45d/update.go#L111-L114

Reply:
Yeah, not too familiar with runc codebase. I only checked with the definition section https://github.com/kubernetes/kubernetes/blame/master/vendor/github.com/opencontainers/runc/libcontainer/configs/cgroup_linux.go#L118, but if it's not wired throughout, then we don't need to expose it here.

Just out of curiosity, is there any reason/historical context why swappiness is defined but not wired?

Reply:
Checked with @haircommander---it's wired via the config that you mentioned, but not exposed as a CLI flag. It is part of the runtime spec: https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/specs-go/config.go#L308

However, I think I'd still lean towards not including this, because none of our user stories have a use case for setting swappiness right now, and it can be configured at the host level.

Reply:
For swap tuning, user need to finetune both `swappiness` and `swap_limit`, my intention is to support both parameters so that user can have all the relevant knobs through kubelet config instead of checking both the kubelet config as well as vm setting. 

That said, I don't oppose dropping swappiness from this KEP if there's no concrete use case ask.